File: leetcode/problems/problem-specific/39.combination-sum/39.combination-sum.go
package main

func combinationSum(candidates []int, target int) [][]int {
	if len(candidates) <= 0 {
		return [][]int{}
	}

	res := [][]int{}
	combinationSumHelper(candidates, target, 0, []int{}, &res)
	return res
}

func combinationSumHelper(candidates []int, target, start int, combination []int, res *[][]int) {
	if target < 0 {
		return
	}
	if target == 0 {
		*res = append(*res, append([]int{}, combination...))
		return
	}

	for i := start; i < len(candidates); i++ {
		combination = append(combination, candidates[i])
		combinationSumHelper(candidates, target-candidates[i], i, combination, res)
		combination = combination[:len(combination)-1]
	}
}

File: leetcode/problems/problem-specific/39.combination-sum/39.combination-sum.py
#
# @lc app=leetcode id=39 lang=python3
#
# [39] Combination Sum
#

# @lc code=start

# * Backtracking Solution | O(n*(2^t)) Time | O(n) Space
# * n -> The length of candidates array | t -> Given target

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []

        def combinationSumHelper(start=0, target=target, combination=[]):
            if target < 0:
                return
            if target == 0:
                result.append(combination[:])
                return

            for i in range(start, len(candidates)):
                combination.append(candidates[i])
                combinationSumHelper(i, target - candidates[i], combination)
                combination.pop()

        combinationSumHelper()
        return result

# @lc code=end
